baybe/utils/sampling_algorithms.py

```diff
@@ -70,3 +76,46 @@ def farthest_point_sampling(
         remaining_point_indices.remove(selected_point_index)
 
     return selected_point_indices
+
+
+class SamplingMethod(Enum):
+    """Available sampling methods for dataframes."""
+
+    Random = "Random"
+    """Random Sampling."""
+
+    FPS = "FPS"
+    """Farthest point sampling."""
+
+
+def sample_numerical_df(
+    df: pd.DataFrame, n_points: int, method: SamplingMethod
+) -> pd.DataFrame:
+    """Sample data points from a data frame.
+
+    Args:
+        df: Data frame with purely numerical entries.
+        n_points: Number of points to sample.
+        method: Sampling method.
+
+    Returns:
+        The sampled points.
+
+    Raises:
+        TypeError: If df has non-numerical content.
+        ValueError: When an invalid sampling method was provided.
+    """
+    if any(df[col].dtype.kind not in "iufb" for col in df.columns):
+        raise TypeError(
+            "'sample_numerical_df' only supports purely numerical data frames."
+        )
+
+    if method is SamplingMethod.FPS:
```

Comment:
Sampling happens without replacement in both cases (which is good, I think). But we have no mechanism yet that catches the case where the number of requested points is larger than the df size – neither for random nor as part of the fps method. I think we should somehow handle it?

Reply:
if added replacement to the utility function and also tests

replacement is an option now available, but it only affects random sampling, it doenst make sense for clustered sampling imo

the case where too many points are requested is now handled too

Reply:
I was first surprised when I saw how you handled the replacement (i.e. instead of throwing an error if `n_points` is larger than the df size and `replacement=False`, you "override" `replacement in a sense) but I see where you are going and I think the rationale is OK here, since it automatically avoids our recommenders to fail. However, I think it's a bit inconsistent how the urn augmentation is done in both cases, and I see issues in both approaches:
* For FPS, you effectively lose the FPS character completely when `n_points>len(df)`, i.e. the oversampling uses a completely different mechanism than the undersampling
* For Random, the mechanism is different from FPS in the overcomplete case in the sense that FPS ensures all points are in the list at least once but random doesn't

I suggest to solve issues by changing the approach slightly:
* For undersampling, we use the original approaches in both cases
* For oversampling, we simply add a multiple of the entire point set first so that we can simply use the undersampling approach on the remainder. That is, if `len(df)=100` and we request `n_points=314`, then we can just add three times the entire set and add the result on top that we get from the original algorithm with `n_points=14`.

What do you think, does it make sense to you? That would also make the code easier in that the first step of the oversampling can be handled upfront, in the same way for all methods. Of course, this should then be explained in the docstring accordingly, since is this is not "classical" replacement – but it works regardless of the selected algorithm.

Reply:
uff im not sure that makes sense. It would simply repeat the undersampeld FPS trajectory for the oversampled part? For a nonprobabalistic method like FPS theres simply no reasonable way to do things when n_requested is larger than n_points, eg I dont see how the method makes more sense than the implemented one?

I would be careful even with aiming to make the two methods too similar, FPS is not a probabalistic method, treating it with the `urnes and replace=on/off` formalism doesnt make sense

IF we want the mechanisms to become more similar then we should change the random sampling to have always `replace=False` for the undersampled part, and then treat the oversampled part the same as FPS currently. In this scenario `replace` could be dropped entirely, it is not needed anywhere in any case. The code would also become simpler in this version

Just throwing an error is also an option

Reply:
changed as discussed